### ISSUE TITLE
Prise en charge des PDFs (extraction des datas) et intégration dans la database.

### ISSUE THREAD

**Issue by jeanbaptisteassouad** [ArtemisMucaj/gearbox-website#14]

**Comment by jeanbaptisteassouad** [comment]
Utilisation du deep learning ?

Python : Keras
http://keras.io/

Video introduction deep learning en python avec Theano (Keras l'utilise en arrière plan) : 
https://www.youtube.com/watch?v=S75EdAcXHKk

Wikipedia deep learning : 
https://fr.wikipedia.org/wiki/Deep_learning

Wikipedia books machin learning : 
https://en.wikipedia.org/wiki/Book:Machine_Learning_%E2%80%93_The_Complete_Guide

**Comment by jeanbaptisteassouad** [comment]
https://www.quora.com/How-do-you-decide-which-loss-function-to-use-for-machine-learning

**Comment by jeanbaptisteassouad** [comment]
http://neuralnetworksanddeeplearning.com/index.html

**Comment by jeanbaptisteassouad** [comment]
http://stats.stackexchange.com/questions/181/how-to-choose-the-number-of-hidden-layers-and-nodes-in-a-feedforward-neural-netw

**Comment by jeanbaptisteassouad** [comment]
Tutorial de stanford : http://ufldl.stanford.edu/tutorial/

**Comment by jeanbaptisteassouad** [comment]
https://www.quora.com/How-does-deep-learning-work-and-how-is-it-different-from-normal-neural-networks-and-or-SVM

**Comment by jeanbaptisteassouad** [comment]
video : https://www.youtube.com/watch?v=ZmNOAtZIgIk

**Comment by jeanbaptisteassouad** [comment]
Un autre truc en python : http://scikit-learn.org/stable/index.html
Ya plein de ressources : http://scikit-learn.org/stable/unsupervised_learning.html

**Comment by jeanbaptisteassouad** [comment]
In the neural network terminology:

one epoch = one forward pass and one backward pass of all the training examples
batch size = the number of training examples in one forward/backward pass. The higher the batch size, the more memory space you'll need.
number of iterations = number of passes, each pass using [batch size] number of examples. To be clear, one pass = one forward pass + one backward pass (we do not count the forward pass and backward pass as two different passes).

Example: if you have 1000 training examples, and your batch size is 500, then it will take 2 iterations to complete 1 epoch.


http://stats.stackexchange.com/questions/153531/what-is-batch-size-in-neural-network

**Comment by jeanbaptisteassouad** [comment]
Pour convertir pdf to png : pdftoppm -rx 300 -ry 300 -png HPC-T4-2013-GearsAndSprockets-GB.pdf prefix

installer poppler-utils sous linux

http://askubuntu.com/questions/50170/how-to-convert-pdf-to-image/50180#50180

**Comment by ArtemisMucaj** [comment]
http://opencv-python-tutroals.readthedocs.org/en/latest/py_tutorials/py_imgproc/py_houghlines/py_houghlines.html

**Comment by jeanbaptisteassouad** [comment]
Documentation de PIL, traitement d'image simple pour python : 
http://effbot.org/zone/pil-index.htm

**Comment by jeanbaptisteassouad** [comment]
Test ce code sur une image avec un tableau : 

from PIL import Image
from PIL import ImageFilter
import numpy as np
import matplotlib.pyplot as plt

def main():
    image = Image.open("prefix-214.png")
    image = image.convert("L")
    image = image.filter(ImageFilter.FIND_EDGES)
    # On convert l'image en array 2D
    width, height = image.size
    data = np.array(image.getdata())
    matrice = np.zeros((height, width))
    for i in xrange(0, height):
        for j in xrange(0, width):
            matrice[i][j] = data[i*width+j]
            pass
        pass

    # On sommes les colonnes et les lignes (on ne prend pas en compte les bords)
    projLignes = np.zeros(height)
    projColonnes = np.zeros(width)
    for i in xrange(1, height-1):
        for j in xrange(1, width-1):
            projLignes[i] += matrice[i][j]
            projColonnes[j] += matrice[i][j]
            pass
        pass
    #print projLignes
    #print projColonnes
    plt.plot(projLignes)
    #plt.show()

    # On fais une moyenne mobile sur projLignes
    tailleFenetre = 100
    projLignesMM = np.zeros(len(projLignes))
    for x in xrange(tailleFenetre, len(projLignes)-tailleFenetre):
        for y in xrange(-tailleFenetre, tailleFenetre+1):
            projLignesMM[x] += projLignes[x+y]
            pass
        pass
    coef = max(projLignes)/max(projLignesMM)
    for x in xrange(0, len(projLignesMM)):
        projLignesMM[x] *= coef
        pass
    plt.plot(projLignesMM)
    plt.show()

    #image.show()
    pass

if __name__ == '__main__':
    main()

**Comment by jeanbaptisteassouad** [comment]
![capture d ecran 2015-11-01 a 18 59 38](https://cloud.githubusercontent.com/assets/10159486/10870424/c8b61d86-80ca-11e5-9012-c25f8a439a5b.png)

**Comment by ArtemisMucaj** [comment]
http://cs231n.github.io/ vraiment bien foutu

**Comment by ArtemisMucaj** [comment]
![table_detection](https://cloud.githubusercontent.com/assets/3483139/10897589/13d684ba-81c3-11e5-9a02-e8e807f9f1c9.png)

ça marche plutôt pas mal comme ça, par contre j'ai du faire un trick relou et utiliser deux fois l'algo de hough pck il voulais pas detecter les horizontales et verticales en mm temps, du coup j'ai séparé mais c'est surement du au fait que je comprends pas forcement exactement l'utilisation des paramètres (peut être ?)

**Comment by jeanbaptisteassouad** [comment]
from PIL import Image
from PIL import ImageFilter
import numpy as np
import matplotlib.pyplot as plt

def main():
    image = Image.open("prefix-034.png")
    image = image.convert("L")
    image = image.filter(ImageFilter.FIND_EDGES)
    # On convert l'image en array 2D
    width, height = image.size
    data = np.array(image.getdata())
    matrice = np.zeros((height, width))
    for i in xrange(0, height):
        for j in xrange(0, width):
            matrice[i][j] = data[i*width+j]
            pass
        pass

    # On sommes les colonnes et les lignes (on ne prend pas en compte les bords)
    projLignes = np.zeros(height)
    projColonnes = np.zeros(width)
    for i in xrange(1, height-1):
        for j in xrange(1, width-1):
            projLignes[i] += matrice[i][j]
            projColonnes[j] += matrice[i][j]
            pass
        pass
    #print projLignes
    #print projColonnes
    for x in xrange(0, len(projLignes)):
        projLignes[x] *= 100.0 / (255*(width-2) )
        pass
    plt.plot(projLignes)
    #plt.show()


    image = Image.open("prefix-006.png")
    image = image.convert("L")
    image = image.filter(ImageFilter.FIND_EDGES)
    # On convert l'image en array 2D
    width, height = image.size
    data = np.array(image.getdata())
    matrice = np.zeros((height, width))
    for i in xrange(0, height):
        for j in xrange(0, width):
            matrice[i][j] = data[i*width+j]
            pass
        pass

    # On sommes les colonnes et les lignes (on ne prend pas en compte les bords)
    projLignes = np.zeros(height)
    projColonnes = np.zeros(width)
    for i in xrange(1, height-1):
        for j in xrange(1, width-1):
            projLignes[i] += matrice[i][j]
            projColonnes[j] += matrice[i][j]
            pass
        pass
    #print projLignes
    #print projColonnes
    for x in xrange(0, len(projLignes)):
        projLignes[x] *= 100.0 / (255*(width-2) )
        pass
    plt.plot(projLignes)
    #plt.show()


    # # On fais une moyenne mobile sur projLignes
    # tailleFenetre = 100
    # projLignesMM = np.zeros(len(projLignes))
    # for x in xrange(tailleFenetre, len(projLignes)-tailleFenetre):
    #     for y in xrange(-tailleFenetre, tailleFenetre+1):
    #         projLignesMM[x] += projLignes[x+y]
    #         pass
    #     pass
    # coef = max(projLignes)/max(projLignesMM)
    # for x in xrange(0, len(projLignesMM)):
    #     projLignesMM[x] *= coef
    #     pass
    # plt.plot(projLignesMM)

    # # Nouvelle projection test
    # tmp = 0
    # projLignes2 = np.zeros(height)
    # projColonnes2 = np.zeros(width)
    # incrTmp = 0.001
    # for i in xrange(1, height-1):
    #     for j in xrange(1, width-1):
    #         if matrice[i][j] == 0:
    #             tmp = 0
    #         else:
    #             if tmp < 1:
    #                 tmp += incrTmp
    #                 pass
    #         pass
    #         projLignes2[i] += matrice[i][j]*tmp
    #         projColonnes2[j] += matrice[i][j]*tmp
    #     pass
    # maxWidthProj2 = 0
    # tmp = 0
    # for x in xrange(1,width-1):
    #     if tmp < 1:
    #         tmp += incrTmp
    #         pass
    #     maxWidthProj2 += 255*tmp
    #     pass
    # # coef = max(projLignes)/max(projLignes2)
    # # for x in xrange(0, len(projLignes2)):
    # #     projLignes2[x] *= coef
    # #     pass
    # for x in xrange(0, len(projLignes2)):
    #     projLignes2[x] *= 100.0 / maxWidthProj2
    #     pass
    # plt.plot(projLignes2)



    plt.show()

    #image.show()
    pass

if __name__ == '__main__':
    main()

**Comment by jeanbaptisteassouad** [comment]
Réduction dimensions image http://arxiv.org/pdf/1304.2683.pdf
